src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -154,89 +210,159 @@ export class MongoDBEventStore implements EventStore<number> {
       });
     }
 
-    // TODO: better error here, should rarely happen if ever
-    // if another error was not thrown before this
+    // TODO: error handling / implement retries
     if (!stream) throw new Error('Failed to create stream');
 
     assertExpectedVersionMatchesCurrent(
-      stream.events.length,
+      stream.metadata.streamPosition,
       options?.expectedStreamVersion,
       MongoDBEventStoreDefaultStreamVersion,
     );
 
+    const updates: UpdateFilter<EventStream> = {
+      $push: { events: { $each: eventCreateInputs } },
+      $set: {
+        'metadata.updatedAt': new Date(),
+        'metadata.streamPosition':
+          stream.metadata.streamPosition + BigInt(events.length),
+      },
+    };
+
+    if (this.asyncProjections) {
+      // Pre-commit handle async projections
+      await handleAsyncProjections({
+        streamName,
+        events: eventCreateInputs,
+        projections: this.asyncProjections,
+        collection: this.collection,
+      });
+    }
+
+    if (this.inlineProjections) {
+      const projections = filterProjections(
+        this.inlineProjections,
+        eventCreateInputs,
+      );
+      for (const { name, handle } of projections) {
+        updates.$set![`projections.${name}`] = await handle(eventCreateInputs, {
+          streamName,
+          collection: this.collection,
+        });
+      }
+    }
+
     // @ts-expect-error The actual `EventType` is different across each stream document,
     // but the collection was instantiated as being `EventStream<Event>`. Unlike `findOne`,
     // `findOneAndUpdate` does not allow a generic to override what the return type is.
     const updatedStream: WithId<EventStream<EventType>> | null =
       await this.collection.findOneAndUpdate(
         {
           streamName: { $eq: streamName },
-          events: { $size: stream.events.length },
+          'metadata.streamPosition': stream.metadata.streamPosition,
         },
-        {
-          $push: { events: { $each: eventCreateInputs } },
-          $set: { updatedAt: new Date() },
-        },
-        { returnDocument: 'after' },
+        updates,
+        { returnDocument: 'after', useBigInt64: true },
       );
 
     if (!updatedStream) {
-      const currentStream = await this.collection.findOne({
-        streamName: { $eq: streamName },
-      });
+      const currentStream = await this.collection.findOne(
```

Comment:
**CONCERN:** I'd prefer not to have the actual position here in error and not need to read the whole document again. Especially that's not a precise information, as someone else could make the update in the  meantime.